pkg/approver/approver.go

```diff
@@ -0,0 +1,243 @@
+package approver
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+
+	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+
+	"github.com/openshift/library-go/pkg/config/clusteroperator/v1helpers"
+	"github.com/openshift/library-go/pkg/operator/status"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	clusterOperatorName           = "machine-approver"
+	clusterOperatorNamespace      = "openshift-cluster-machine-approver"
+	unknownVersionValue           = "unknown"
+	operatorVersionKey            = "operator"
+	reasonAsExpected              = "AsExpected"
+	releaseVersionEnvVariableName = "RELEASE_VERSION"
+)
+
+var relatedObjects = []configv1.ObjectReference{
+	{
+		Group:    "",
+		Resource: "namespaces",
+		Name:     clusterOperatorNamespace,
+	},
+	{
+		Group:    "certificates.k8s.io",
+		Resource: "certificatesigningrequests",
+		Name:     "",
+	},
+}
+
+type StatusController struct {
+	client.Client
+	status.VersionGetter
+}
+
+func (c *StatusController) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	controller, err := ctrl.NewControllerManagedBy(mgr).
+		WithOptions(options).
+		For(&configv1.ClusterOperator{}).
+		WithEventFilter(predicate.Funcs{
+			CreateFunc:  func(e event.CreateEvent) bool { return clusterOperatorFilter(e.Object) },
+			UpdateFunc:  func(e event.UpdateEvent) bool { return clusterOperatorFilter(e.ObjectNew) },
+			DeleteFunc:  func(e event.DeleteEvent) bool { return clusterOperatorFilter(e.Object) },
+			GenericFunc: func(e event.GenericEvent) bool { return clusterOperatorFilter(e.Object) },
+		}).
+		Build(c)
+	if err != nil {
+		return err
+	}
+
+	// Watch operator controller version changes
+	if err = controller.Watch(
+		&source.Channel{Source: c.NewVersionWatcher()},
+		&handler.EnqueueRequestForObject{},
+	); err != nil {
+		return err
+	}
```

Comment:
@JoelSpeed The problem is, that in case the operator deployment will change it's `RELEASE_VERSION` after upgrade, the `clusterOperator` could be left unchanged. Or are the upgrades recreating `clusterOperator` manifests from scratch?

Reply:
Oh I think I've misunderstood, the version is coming from another resource is it? In that case sure, this is needed :) 
Might want to make the method private though